pkg/k8s/identitybackend/identity.go

```diff
@@ -0,0 +1,433 @@
+			},
+		},
+	}
+
+	obj, exists, err := c.Store.Get(identityTemplate)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, nil
+	}
+
+	identity, ok := obj.(*types.Identity)
+	if !ok {
+		return nil, fmt.Errorf("invalid object")
+	}
+
+	return c.KeyType.PutKeyFromMap(identity.SecurityLabels), nil
+}
+
+// Release dissociates this node from using the identity bound to key. When an
+// identity has no references it may be garbage collected.
+func (c *crdBackend) Release(ctx context.Context, key allocator.AllocatorKey) (err error) {
+	identity := c.get(ctx, key)
+	if identity == nil {
+		return fmt.Errorf("unable to release identity %s: identity does not exist", key)
+	}
+
+	if _, ok := identity.Status.Nodes[c.NodeName]; !ok {
+		return fmt.Errorf("unable to release identity %s: identity is unused", key)
+	}
+
+	delete(identity.Status.Nodes, c.NodeName)
+
+	capabilities := k8sversion.Capabilities()
+
+	identityOps := c.Client.CiliumV2().CiliumIdentities()
+	if capabilities.Patch {
+		var patch []byte
+		patch, err = json.Marshal([]JSONPatch{
+			{
+				OP:   "delete",
+				Path: "/status/nodes/" + c.NodeName,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		_, err = identityOps.Patch(identity.GetName(), k8sTypes.JSONPatchType, patch, "status")
+		if err == nil {
+			return nil
+		}
+		log.WithError(err).Debug("Error patching status. Continuing update via UpdateStatus")
+		/* fall through and attempt UpdateStatus() or Update() */
+	}
+
+	identityCopy := identity.DeepCopy()
+	if identityCopy.Status.Nodes == nil {
+		return nil
+	}
+
+	if capabilities.UpdateStatus {
+		_, err = identityOps.UpdateStatus(identityCopy.CiliumIdentity)
+		if err == nil {
+			return nil
+		}
+		log.WithError(err).Debug("Error updating status. Continuing update via Update")
+		/* fall through and attempt Update() */
+	}
+
+	_, err = identityOps.Update(identityCopy.CiliumIdentity)
+	return err
+}
+
+func (c *crdBackend) ListAndWatch(handler allocator.CacheMutations, stopChan chan struct{}) {
+	c.Store = cache.NewStore(cache.DeletionHandlingMetaNamespaceKeyFunc)
+	identityInformer := informer.NewInformerWithStore(
+		cache.NewListWatchFromClient(c.Client.CiliumV2().RESTClient(),
+			"ciliumidentities", v1.NamespaceAll, fields.Everything()),
+		&v2.CiliumIdentity{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if identity, ok := obj.(*types.Identity); ok {
+					if id, err := strconv.ParseUint(identity.Name, 10, 64); err == nil {
+						handler.OnAdd(idpool.ID(id), c.KeyType.PutKeyFromMap(identity.SecurityLabels))
+					}
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if identity, ok := newObj.(*types.Identity); ok {
+					if id, err := strconv.ParseUint(identity.Name, 10, 64); err == nil {
+						handler.OnModify(idpool.ID(id), c.KeyType.PutKeyFromMap(identity.SecurityLabels))
+					}
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if identity, ok := obj.(*types.Identity); ok {
```

Comment:
You need to check for `cache.DeletedFinalStateUnknown` object types. Like it is done here:

https://github.com/cilium/cilium/blob/23fd0d7cd4262ee1fc647f5a1078075b1f1f4fb0/daemon/k8s_watcher.go#L417-L430

Reply:
Done. It looks different from what we have elsewhere, so you might want to look at it again @aanm . I can add a `CopyObjToV1NetworkPolicy` style function to make it look similar. Up to you. I kept it simple for now.